### ISSUE TITLE
`GpuDiagnostics` does not warn when incompatible calls

### ISSUE THREAD

**Issue by Guillaume-Helbecque** [chapel-lang/chapel#23625]
The following code does not warn. 

I think it could be useful to make it warn when `stop` but not `start`, or `get` but not `start`, etc. , as it is for `stopwatch` for example.

```chapel
use GpuDiagnostics;

stopGpuDiagnostics();
```

**Compile command:**
`chpl foo.chpl`

**Execution command:**
`./foo`

**Associated Future Test(s):**

### Configuration Information

- Output of `chpl --version`: 
```
chpl version 1.32.0
  built with LLVM version 15.0.7
```
- Output of `$CHPL_HOME/util/printchplenv --anonymize`:
```
CHPL_TARGET_PLATFORM: linux64
CHPL_TARGET_COMPILER: llvm
CHPL_TARGET_ARCH: x86_64
CHPL_TARGET_CPU: native
CHPL_LOCALE_MODEL: gpu *
  CHPL_GPU: nvidia *
CHPL_COMM: none
CHPL_TASKS: qthreads
CHPL_LAUNCHER: none
CHPL_TIMERS: generic
CHPL_UNWIND: none
CHPL_MEM: jemalloc
CHPL_ATOMICS: cstdlib
CHPL_GMP: bundled
CHPL_HWLOC: bundled
CHPL_RE2: bundled
CHPL_LLVM: bundled *
CHPL_AUX_FILESYS: none
```
- Back-end compiler and version, e.g. `gcc --version` or `clang --version`: `gcc (Spack GCC) 12.2.0`

**Comment by e-kayrakli** [comment]
Thanks for the request @Guillaume-Helbecque! I agree that this would be nice to have. GpuDiagnostics is modeled after CommDiagnostics. I suspect the behavior you described here is identical between the two modules, and if so, it'd be good to add this feature to both of those modules.

**Comment by e-kayrakli** [comment]
If you called `startGpuDiagnostics`, but the program execution ends without calling `stopGpuDiagnostics`; should that be a warning? An error?

**Comment by bradcray** [comment]
It seems as though if the program is already existing due to an erroneous condition through a halt/assert/uncaught throw/etc., it would be a shame to print something additional about a started diagnostics not being stopped (in the sense of "I wasn't expecting to exit this way, don't blame me, the user!").  So I'd either probably not make that a warning/error, or else only make it one if the exit status is 0.

**Comment by jabraham17** [comment]
I have implemented two warnings for this in #23948. I added the same warnings to CommDiagnostics and GpuDiagnostics.

I added a warning for start without stop, but also for start after start. Does anyone have any objections to either warning? See the PR for details of the warning.
